library/bignum_core.h

```diff
@@ -0,0 +1,157 @@
+/**
+ *  Core bignum functions
+ *
+ *  This interface only should be used by the legacy bignum module (bignum.h)
+ *  and the modular bignum modules (bignum_mod.c, bignum_mod_raw.c). All other
+ *  modules should use the high level modular bignum interface (bignum_mod.h)
+ *  or the legacy bignum interface (bignum.h).
+ *
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#ifndef MBEDTLS_BIGNUM_CORE_H
+#define MBEDTLS_BIGNUM_CORE_H
+
+#include "common.h"
+
+#if defined(MBEDTLS_BIGNUM_C)
+#include "mbedtls/bignum.h"
+#endif
+
+/** Count leading zero bits in a given integer.
+ *
+ * \param x     Integer to count leading zero bits.
+ *
+ * \return      The number of leading zero bits in \p x.
+ */
+size_t mbedtls_mpi_core_clz( const mbedtls_mpi_uint x );
+
+/** Return the the minimum number of bits required to represent the value held
+ * in the MPI.
+ *
+ * \note This function returns 0 if all the limbs of \p X are 0.
+ *
+ * \param X     The address of the MPI.
+ * \param nx    The number of limbs of \p X.
+ *
+ * \return      The number of bits in \p X.
+ */
+size_t mbedtls_mpi_core_bitlen( const mbedtls_mpi_uint *X, size_t nx );
+
+/** Convert a big-endian byte array aligned to the size of mbedtls_mpi_uint
+ * into the storage form used by mbedtls_mpi.
+ *
+ * \param X     The address of the MPI.
+ * \param limbs The number of limbs of \p X.
+ */
+void mbedtls_mpi_core_bigendian_to_host( mbedtls_mpi_uint * const X,
+                                         size_t limbs );
```

Comment:
minor: All other limb params in this header file use `nx`, this could be changed to follow the same pattern.

Reply:
heh, I think all the ones I've looked at in the monmul PR have used just `n`

@yanesca - please, we need a standard!!

Reply:
This used to be static and was taken from bignum.c, and the rest was taken from the prototype, hence the difference (and of course the prototype isn't always consistent either).

- In `_mod` this information is encapsulated in structures, but there we are using `limbs` and I think it is good that way (https://github.com/Mbed-TLS/mbedtls/pull/6095#discussion_r942387289).
- In `_mod_raw` we are not passing length for operands, the length is taken from the modulus.
- In `_core` we don't have any convention and using all kind of names like `n`, `nx`, `B_len`

I suggest that going forward we use `limbs` if only one length parameter is passed and qualify them with a prefix if there are more (eg `nlimbs` or `blimbs`).

What do you think?

Reply:
Thought about it and for cognitive ease it is probably better to use qualified limb parameters wherever possible (eg. when all the input have the same number of limbs and only one limbs parameter is passed, it would remain unqualified).

Also using a proper separation (`n_limbs` ) helps readability as well.